### PR TITLE
Support k parameter for knn query

### LINKED ISSUE #108473: Support `k` parameter for `knn` query
### Description

Currently `knn` query doesn't support `k` parameter, and `size` parameter of a search request is used as `k`.

But for hybrid search, it could be useful to have a separate `k` parameter to ensure only `k` results from knn search are considered. This `k` parameter will be shard based.

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Introduce an optional k param for knn query

If k is not set, knn query has the previous behaviour:
- `num_candidates` docs  is collected from each shard. This `num_candidates` docs are used for combining with results with other queries and aggregations on each shard.
- docs from all shards are merged to produce the top global `size` results

If k is set, the behaviour instead is following:
- `k` docs is collected from each shard. This `k` docs are used for combining results with other queries and aggregations on each shard.
- similarly, docs from all shards are merged to produce the top global `size` results.

Having `k` param makes it more intuitive for users to address their needs. They also don't need to care and can skip `num_candidates` param for this query as it is of more internal details to tune how knn search operates.

Closes #108473

### CODE REVIEW

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/190_knn_query-with-k-param.yml [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+                field: my_name
+                order:
+                  _key: asc
+
+  - length: {hits.hits: 2}
+  - match: {hits.total.value: 5}
+  - match: {aggregations.my_agg.buckets.0.key: 'v1'}
+  - match: {aggregations.my_agg.buckets.1.key: 'v2'}
+  - match: {aggregations.my_agg.buckets.0.doc_count: 3}
+  - match: {aggregations.my_agg.buckets.1.doc_count: 2}
```
Comment (benwtrent) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/190_knn_query-with-k-param.yml:204]: Could you add some tests for the bad input case? `k=0` and `k < num_candidates`?

**Review by benwtrent — COMMENTED** [review]
❤️ its so simple! I have one comment regarding testing :)

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/190_knn_query-with-k-param.yml (hunk 1 above)
Comment (mayya-sharipova) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/190_knn_query-with-k-param.yml:236]: Thanks for the review, addressed in 7f70a316d4af7585e1b460b26bbf2924ac0a55b4

**Review by benwtrent — APPROVED** [review]

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/part-3

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run this please

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/7.17.23 / bwc-snapshots

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/bwc-snapshots

File: server/src/main/java/org/elasticsearch/search/vectors/ESDiversifyingChildrenFloatKnnVectorQuery.java [hunk 2]
```diff
@@ -15,15 +15,24 @@
...
+        BitSetProducer parentsFilter
+    ) {
+        super(field, query, childFilter, numCands, parentsFilter);
+        this.kParam = k;
     }
 
     @Override
     protected TopDocs mergeLeafResults(TopDocs[] perLeafResults) {
-        TopDocs topK = super.mergeLeafResults(perLeafResults);
+        TopDocs topK = kParam == null ? super.mergeLeafResults(perLeafResults) : TopDocs.merge(kParam, perLeafResults);
```
Comment (pmpailis) [inline server/src/main/java/org/elasticsearch/search/vectors/ESDiversifyingChildrenFloatKnnVectorQuery.java:35]: Not related to this PR, but going to the `super` implementation and seeing that we also have `TopDocs.merge(k, perLeafResults);` where `k == numCands` is a bit confusing.

File: server/src/test/java/org/elasticsearch/search/vectors/AbstractKnnVectorQueryBuilderTestCase.java [hunk 3]
```diff
@@ -163,16 +166,27 @@ public void testWrongFieldType() {
...
             new float[] { 1.0f, 1.0f, 1.0f },
+            5,
             10,
             null
         );
         IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () -> query.doToQuery(context));
         assertThat(e.getMessage(), containsString("[knn] queries are only supported on [dense_vector] fields"));
     }
 
+    public void testKLessThanNumCands() {
```
Comment (pmpailis) [inline server/src/test/java/org/elasticsearch/search/vectors/AbstractKnnVectorQueryBuilderTestCase.java:177]: Was confused a bit by the test name; shouldn't it be the other way around? :D

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/190_knn_query-with-k-param.yml [hunk 4]
```diff
@@ -0,0 +1,236 @@
+# test how knn query interact with other queries
+setup:
```
Comment (pmpailis) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/190_knn_query-with-k-param.yml:2]: Maybe we could also add a test where we have a `knn` query through a `standard` retriever?

**Review by pmpailis — APPROVED** [review]
Nice work @mayya-sharipova ! ❤️

File: server/src/test/java/org/elasticsearch/search/vectors/AbstractKnnVectorQueryBuilderTestCase.java (hunk 3 above)
Comment (mayya-sharipova) [reply server/src/test/java/org/elasticsearch/search/vectors/AbstractKnnVectorQueryBuilderTestCase.java]: @pmpailis Indeed, addressed in b5a189af78b9009f538bfc93662fa4ec8ae0e124

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/190_knn_query-with-k-param.yml (hunk 4 above)
Comment (mayya-sharipova) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/190_knn_query-with-k-param.yml:2]: +1, added test in b5a189af78b9009f538bfc93662fa4ec8ae0e124

File: server/src/main/java/org/elasticsearch/search/vectors/ESDiversifyingChildrenFloatKnnVectorQuery.java (hunk 2 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/search/vectors/ESDiversifyingChildrenFloatKnnVectorQuery.java:35]: Yes, indeed. You are right.

**Review by tteofili — APPROVED** [review]

**Pull request merged**